src/isa/instruction.h

```diff
@@ -393,6 +393,33 @@ struct ImmPartsImpl : public ImmParts... {
   constexpr static void Decode(Instr_T &value, const Instr_T instruction) {
     (ImmParts::Decode(value, instruction), ...);
   }
+
+private:
+  template <typename FirstPart, typename... OtherParts>
+  struct Verify {};
+  template <typename FirstPart, typename SecondPart, typename... OtherParts>
+  struct Verify<FirstPart, SecondPart, OtherParts...> {
+    /// Returns true if FirstPart and SecondPart are not overlapping
+    constexpr static bool IsNotOverlapping() {
+      return (FirstPart::Offset() >
+                  (SecondPart::Offset() + SecondPart::BitRange::Width()) ||
+              SecondPart::Offset() >
+                  (FirstPart::Offset() + FirstPart::BitRange::Width()));
+    }
+    enum {
+      nonOverlapping = (IsNotOverlapping() &&
+                        Verify<FirstPart, OtherParts...>::nonOverlapping &&
+                        Verify<SecondPart, OtherParts...>::nonOverlapping)
+    };
+  };
```

Comment:
Very clever!

Reply:
Thanks, I'm really proud of those verification structures! I have a few more ideas on other parts of the class structure that can be similarly verified (ex: verify that every bit is accounted for), but do you think I should focus on re-implementing the ISA first? The benefit of adding more verifications first is that when I re-implement the instructions I will be sure that they are verified to be valid.

Reply:
>  but do you think I should focus on re-implementing the ISA first?

Your call - if you can get it done in a reasonable amount of time and you estimate that time to be less than the time you'll need for debugging this - sure. But you can also always create issues to capture these things and get back to it later.